File: 238-product-of-array-except-self/product-of-array-except-self.cpp
class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        
        int n = nums.size();
        vector<int>op(n, 0);
        // approach 1 : for each index, find prod of elem before i, find prod of elem after i, and store -> O(n^2) approach

        // approach 2 : to optimize this, store the elements in an array, then just compute -> use prefix and suffix

        // approach 3: no need to store prefix, compute prefix on the go, just store suffix, that will be our output array as well

        int prod = 1;
        for(int i = n-1; i >=0 ; i--)
        {
            op[i] = prod;
            prod*=nums[i];
        }
        prod = 1;
        for(int i = 0 ; i < n; i++)
        {
            op[i] = op[i]*prod;
            prod*=nums[i];
        }
        return op;
    }
};

File: 238-product-of-array-except-self/product-of-array-except-self.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        # 2 loops-> O(n^2) time complexity, O(1) space
        # for every element, store the product of elements on left and right

        # suffix = [1]*len(nums)
        ans = [1]*len(nums)

        prod = 1
        for i in range(len(nums)-1, -1, -1):
            ans[i] = prod
            prod*=nums[i]
        
        prod = 1
        for i in range(len(nums)):
            ans[i] = prod*ans[i]
            prod = prod*nums[i]

        return ans
        
